src/JuMP.jl

```diff
@@ -335,21 +335,126 @@ Base.broadcastable(model::Model) = Ref(model)
     backend(model::Model)
 
 Return the lower-level MathOptInterface model that sits underneath JuMP. This
-model depends on which operating mode JuMP is in (see [`mode`](@ref)), and
-whether there are any bridges in the model.
+model depends on which operating mode JuMP is in (see [`mode`](@ref)).
 
-If JuMP is in `DIRECT` mode (i.e., the model was created using
-[`direct_model`](@ref)), the backend will be the optimizer passed to
-[`direct_model`](@ref).
+ * If JuMP is in `DIRECT` mode (i.e., the model was created using
+   [`direct_model`](@ref)), the backend will be the optimizer passed to
+   [`direct_model`](@ref).
+ * If JuMP is in `MANUAL` or `AUTOMATIC` mode, the backend is a
+   `MOI.Utilities.CachingOptimizer`.
 
-If JuMP is in `MANUAL` or `AUTOMATIC` mode, the backend is a
-`MOI.Utilities.CachingOptimizer`.
+**This function should only be used by advanced users looking to access
+low-level MathOptInterface or solver-specific functionality.**
 
-This function should only be used by advanced users looking to access low-level
-MathOptInterface or solver-specific functionality.
+## Notes
+
+If JuMP is not in `DIRECT` mode, the type returned by `backend` may change
+between any JuMP releases. Therefore, only use the public API exposed by
+MathOptInterface, and do not access internal fields such as
+`backend(model).optimizer.model`. If you require access to the
+innermost optimizer, see [`unsafe_backend`](@ref). Alternatively, use
+[`direct_model`](@ref) to create a JuMP model in `DIRECT` mode.
+
+See also: [`unsafe_backend`](@ref).
 """
 backend(model::Model) = model.moi_backend
 
+"""
+    unsafe_backend(model::Model; attach_caching_optimizers::Bool = true)
+
+Return the innermost optimizer associated with the JuMP model `model`, skipping
+`CachingOptimizer` and `LazyBridgeOptimizer` layers as needed.
+
+If `attach_caching_optimizers`, `MOIU.attach_optimizer(model)` will be called on
+all `MOIU.CachingOptimizer`s in the stack.
+
+**This function should only be used by advanced users looking to access
+low-level solver-specific functionality. It has a high-risk of incorrect usage.
+We strongly suggest you use the alternative suggested below.**
+
+See also: [`backend`](@ref).
+
+## Unsafe behavior
+
+This function is unsafe because the innermost optimizer may not represent the
+same problem as `model`.
+
+If you modify the JuMP model, the reference you have to the backend may be
+invalid. Using an invalid reference may result in incorrect solutions, errors,
```

Comment:
What makes it actually invalid? It is just a dangling reference to an object that will be freed by the GC, or do we explicitly invalidate the models (e.g., free licenses)?

Reply:
The Julia `Optimizer` object will still be the same, we don't invalidate the reference, we just call `MOI.empty!`

Reply:
So the backend (the Optimizer) won't be invalid itself, but everything obtained from it might be, right?

Reply:
Invalid might be the wrong word. I've changed to `out-dated`. 

The `Optimizer` object is still the same, but the inner pointer is probably changed or set to `C_NULL` on `MOI.empty!`, so calling low-level functions could do anything. 

We don't have any API guarantees on how solvers should handle the C models and when they are initialized and freed, only that accessing it through MOI is safe.

Reply:
So we might need to highlight that the advanced user should have some knowledge of the Solver's MOI wrapper.